lib/Sema/SemaExpr.cpp

```diff
@@ -11949,14 +11949,67 @@ ExprResult Sema::ActOnCountBoundsExpr(SourceLocation BoundsKWLoc,
       << ResultType;
     return ExprError();
   }
-  return new (Context) CountBoundsExpr(Kind, PromotedCountExpr, BoundsKWLoc,
+  return new (Context) CountBoundsExpr(Kind, CountExpr, BoundsKWLoc,
                                        RParenLoc);
 }
 
+/// \brief Validate the type of an argument expression to a bounds
+/// expression. If valid, return the unqualified pointee type.  Otherwise
+/// return null.
+const Type *Sema::ValidateBoundsExprArgument(Expr *Arg) {
+  QualType ArgType = Arg->getType();
+  const Type *ArgPointerType = ArgType->getAs<PointerType>();
```

Comment:
Good question.  The function assumes that array->pointer conversion has already been done, so it does not rule out array types.   I've added a comment to the header for the function noting the assumption.